userapi/src/main/java/fk/prof/userapi/api/ProfileStoreAPIImpl.java

```diff
@@ -150,88 +148,80 @@ public void getProfilesInTimeWindow(Future<List<AggregatedProfileNamingStrategy>
-                        result -> completeAggregatedProfileLoading(cache, result, fileNameKey));
-            }
-        } else {
-            future.complete(cachedProfileInfo);
-        }
+    public void load(Future<AggregatedProfileInfo> future, AggregatedProfileNamingStrategy filename) {
+        clusterAwareCache.getAggregatedProfile(filename).setHandler(future.completer());
     }
 
-    @Override
-    synchronized public void loadSummary(Future<AggregationWindowSummary> future, AggregatedProfileNamingStrategy filename) {
-
-        if (!filename.isSummaryFile) {
+    public void loadSummary(Future<AggregationWindowSummary> future, AggregatedProfileNamingStrategy filename) {
+        if(!filename.isSummaryFile) {
             future.fail(new IllegalArgumentException(filename.getFileName(0) + " is not a summaryFile"));
             return;
         }
 
         String fileNameKey = filename.getFileName(0);
-
         AggregationWindowSummary cachedProfileInfo = summaryCache.getIfPresent(fileNameKey);
-        if (cachedProfileInfo == null) {
-            boolean fileLoadInProgress = futuresForLoadingFiles.containsKey(fileNameKey);
-            // save the future, so that it can be notified when the loading visit finishes
-            saveRequestedFuture(fileNameKey, future);
+        if(cachedProfileInfo == null) {
+            boolean fileLoadInProgress = filesBeingLoaded.containsKey(fileNameKey);
+            saveRequestedFutureInMap(fileNameKey, future, filesBeingLoaded);
+
             // set the timeout for this future
-            vertx.setTimer(profileLoadTimeout, timerId -> timeoutRequestedFuture(fileNameKey, future));
+            vertx.setTimer(loadTimeout, timerId -> timeoutRequestedFutureInMap(fileNameKey, future, filesBeingLoaded));
 
-            if (!fileLoadInProgress) {
+            if(!fileLoadInProgress) {
                 workerExecutor.executeBlocking((Future<AggregationWindowSummary> f) -> profileLoader.loadSummary(f, filename),
-                        true,
-                        result -> completeAggregatedProfileLoading(summaryCache, result, fileNameKey));
+                    true,
+                    result -> {
+                        if(result.succeeded()) {
+                            summaryCache.put(fileNameKey, result.result());
+                        }
+                        filesBeingLoaded.remove(fileNameKey).complete(result.map(r -> r));
+                    });
             }
         } else {
             future.complete(cachedProfileInfo);
         }
     }
 
+    @Override
+    public <T extends ProfileView> Future<Pair<AggregatedSamplesPerTraceCtx, T>> getProfileView(AggregatedProfileNamingStrategy profileName, String traceName, ProfileViewType profileViewType) {
+        Future<Pair<AggregatedSamplesPerTraceCtx, T>> result = Future.future();
+
+        Future<Pair<AggregatedSamplesPerTraceCtx, T>> f = clusterAwareCache.getProfileView(profileName, traceName, profileViewType);
```

Comment:
Instead of cache.get -> cache.load -> cache.get (seems counterintuitive), support for cache.getOrLoad can simplify the flow here. getOrLoad (or a better named method) can return a datastructure which has locally cached profileview or the metadata (location of remote profile). It will transparently load profile if not present in cluster

Reply:
I have not seen ClusterAwareCache class yet, but needless to say it should take care of race conditions between get and load. For example load of a profile happening on simultaneous boxes because get returned null. Will check for the same when I review that class

Reply:
Agreed. Made changed to include this.